### ISSUE TITLE
Collision Detection

### ISSUE THREAD

**Issue by izawan** [uhmlavalab/SIEGE#5]
Unsure on how to code the collision detection here. Obviously going to use the getHitBox() method for individual entities. Check for overlap between objects, then call collisionEffects()? The way I coded the collisionEffects function is for it to take two objects and then determine the types, then go about with the effects. Not sure if this is optimal or there is a better way of doing this.

**Comment by d-k-k** [comment]
Edit: Ogad, the formatting was bad.
Sounds fine to me.
But where is collisionEffects() located? 
If it is part of an entity, then it may need only 1 param(  the other object that it collided with)

Some pseudocode example:

``` javascript
for (each obj1 in playerbullets) {
    for(each obj2 in invaders) {
        var hb1 = obj1.getHitBox();
        var hb2 = obj2.getHitBox();
        if(   isTouching(hb1, hb2) ) {
            obj1.collisionEffect(obj2);
            obj2.collisionEffect(obj1);
        }
    }//end for each invader
}//end for each player bullet
```

**Comment by izawan** [comment]
So essentially on the first loop through the outer loop, the first of the three shots available (for each player) will check with every single invader to see if there is an overlap? Just based on the pseudocode anyway. My initial idea was to see if a player's hitbox overlaps with any other object, and if it does, then call the collisionEffects to see if there should be an effect or not, based on the type. Is that not easier? Player bullets can also collide with other players as well as the boss, so I would need to check for all of that too.

**Comment by d-k-k** [comment]
That works too, just need to have an array of everything.
If you do make an array of everything make sure that you have a check to continue if they refer to the same object

``` javascript
for( each obj1 in allObjects) {
  for(each obj2 in allObjects) {
    if(obj1 === obj2) { continue; } //an object is always occupying its own space
    if( isTouching(obj1,obj2) ) {  determineCollisionEffects(obj1, obj2); }
  }//end inner
}//end outer
```

**Comment by izawan** [comment]
Okay, that sounds good. I think how I'll do it is inside of the playerUpdate/invaderUpdate functions I'll include an update that gets the player/invader's hitbox and checks it against all other objects. With that said, would it make that much of a difference if I checked a specific object with all that it is supposed to collide with? For example, if I checked an invader's hitbox with all players and playerBullets, but not other invaders/invaderBullets. It seems a bit inefficient to me to be checking for other invaders and invaderBullets if those things aren't supposed to cause an effect. It'd require making multiple arrays though.
